src/MCPClient/tests/test_archivematicaClient.py

```diff
@@ -0,0 +1,45 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import unicode_literals
+
+from archivematicaClient import handle_batch_task
+
+
+def test_handle_batch_task_replaces_non_ascii_arguments(mocker):
+    # We are only interested in verifying the string replacement logic
+    # for task arguments and mock the remaining functionality
+    mocker.patch("archivematicaClient.auto_close_db")
+    mocker.patch("archivematicaClient.Job")
+    mocker.patch("archivematicaClient.Task")
+    mocker.patch("archivematicaClient.retryOnFailure")
+    mocker.patch(
+        "cPickle.loads",
+        return_value={
+            "tasks": {
+                "some_task_uuid": {
+                    "uuid": "some_task_uuid",
+                    "arguments": "montréal %taskUUID% %jobCreatedDate%",
+                    "createdDate": "some montréal datetime",
+                    "wants_output": False,
+                }
+            }
+        },
+    )
+
+    # The mocked module will not have a `concurrent_instances` attribute
+    mocker.patch(
+        "importlib.import_module", return_value=mocker.MagicMock(spec=["call"])
+    )
+
+    # This is the only function that uses the arguments after the replacements
+    _parse_command_line = mocker.patch("archivematicaClient._parse_command_line")
+
+    # Mock the two parameters sent to handle_batch_task
+    gearman_job_mock = mocker.Mock()
+    supported_modules_mock = mocker.Mock(**{"get.side_effect": "some_module_name"})
+    handle_batch_task(gearman_job_mock, supported_modules_mock)
```

Comment:
I think I understand why this might fail, but not why the inconsistency across environments. What I think I might be able to do is show that the mock on line `11` doesn't work as anticipated though. If you navigate to:

* https://github.com/artefactual/archivematica/blob/6946cb55158ca3fe33e2504c358ace79b1be1ed5/src/archivematicaCommon/lib/databaseFunctions.py#L39-L49 

And stick `assert False` before `close_old_connections()` you'll see that when the test fails, it will have called `handle_batch_task` and the `finally` clause is still hit. Whereas, I think we're trying to avoid any interaction with decorator entirely. We'd like the `handle_batch_task` function unwrapped to test it and not hit `close_old_connections()`. So I think that this is something we might look at fixing? But it's complicated. 

For completeness, this is what I see, when this happens is the transaction error as mentioned in Slack:
```
self = <django.db.backends.mysql.base.DatabaseWrapper object at 0x7f8309366d10>

    def get_autocommit(self):
        """
        Check the autocommit state.
        """
>       self.ensure_connection()
E       RuntimeError: Database access not allowed, use the "django_db" mark, or the "db" or "transactional_db" fixtures to enable it.
```

But only when I run all of the MCP Client tests. Not when I run this one test on its own. 

When I run tox in a clean virtual environment, I can see the finally being called, but I don't receive a tox/pytest warning at all. It seems to interact with Django differently? (NB. this is the same as if I run `sudo docker-compose run --workdir /src/MCPClient --rm --entrypoint=py.test archivematica-mcp-client -p no:cacheprovider -p no:warnings --reuse-db -v tests/test_archivematicaClient.py`) -- I've tried looking at `pytest` versions briefly here, and can see there may be platform differences. I don't know if these are affecting the output.
```
>       handle_batch_task(gearman_job_mock, supported_modules_mock)

src/MCPClient/tests/test_archivematicaClient.py:57: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

args = (<Mock id='140345534660560'>, <Mock id='140345534660496'>), kwargs = {}

    @wraps(f)
    def wrapper(*args, **kwargs):
        try:
            return f(*args, **kwargs)
        finally:
>           assert False
E           AssertionError
```

I don't know what to do to fix it. We can use the `@pytest.mark.django_db` decorator on the test function. It's technically not incorrect, even though we'd rather mock out any calls to the db. 

Otherwise, I've tried:

* https://stackoverflow.com/a/43213188
* Undecorated: https://pypi.org/project/undecorated/

And neither work.

And this person has some mega insights into decorators and doesn't seem pleased by them:

* https://hynek.me/articles/decorators/

In Python 3 we get a `handle_batch_task.__wrapped__` helper which would solve all of this and we wouldn't need the db mocks at all.

Reply:
Thanks for the investigation on the decorator!

At this point I'm tempted to start over the change to `lib.archivematicaClient.py`. As I wrote in the test I'm only interested in verifying that the string replacement logic doesn't break with non-ascii, and I could achieve the same by extracting only this block to a separate function:

https://github.com/artefactual/archivematica/blob/29a0423a9bc1ab77ebe4ae7bfdc2bdde2099ef05/src/MCPClient/lib/archivematicaClient.py#L128-L129

and just testing that. No decorator, no database, no nothing!

That would still leave us with the discrepancies between environments, etc etc etc. What do you think?

Reply:
I went ahead and just marked the test to have interaction with the database and removed the mocking :disappointed: